File: Python/backend/static/js/leaderboard.js
function createCell(text) {
  const td = document.createElement("td");
  td.textContent = text;
  return td;
}

export function renderLeaderboard(leaderboardData) {
  const tbody = document.getElementById("leaderboard-body");
  tbody.innerHTML = "";
  leaderboardData.forEach((row) => {
    const tr = document.createElement("tr");
    tr.appendChild(createCell(row.make));
    tr.appendChild(createCell(row.model));
    tr.appendChild(createCell(row.year));
    tr.appendChild(createCell(row.engine_hp));
    tr.appendChild(createCell(row.highway_mpg));
    tr.appendChild(createCell(row.efficiency_score));
    tbody.appendChild(tr);
  });
}

File: Python/backend/analytics/leaderboard.py
def compute_leaderboard(filtered_df):
    leaderboard_df = filtered_df[filtered_df["engine_hp"] > 0].copy()
    leaderboard_df["efficiency_score"] = (leaderboard_df["highway_mpg"] / (leaderboard_df["engine_hp"] / 100)).round(2)
    leaderboard_df = leaderboard_df.sort_values("efficiency_score", ascending=False).head(10)
    return [
        {
            "make": row["make"],
            "model": row["model"],
            "year": int(row["year"]),
            "engine_hp": float(row["engine_hp"]),
            "highway_mpg": float(row["highway_mpg"]),
            "efficiency_score": float(row["efficiency_score"]),
        }
        for _, row in leaderboard_df.iterrows()
    ]
